### PR TITLE
Implement MPC

### PR DESCRIPTION
**Pull request by chungmin99**
Includes:
- Gradient-based MPC implementation, and
- MVP for sampling-based MPC (mppi) -- the parameters aren't quite tuned.

Also modifies:
- world/self-collision `RobotFactor` batching logic (minimizing for-loops is critical for MPC JIT time).
- manipulability jacobian shaping logic

### CODE REVIEW

File: examples/03_kin_with_coll.py
Comment (chungmin99) [inline examples/03_kin_with_coll.py:1]: Looks like a minor file shuffle edit for a different example (`kin_with_coll`), maybe this could have been done in a separate PR or done with the original PR itself (with more forethought...).

File: src/jaxmp/coll/_collide_types.py [hunk 1]
```diff
@@ -26,11 +26,30 @@
...
+    )
     direction = direction / (jnp.linalg.norm(direction) + 1e-6)
-    y, z = jnp.array([0, 1, 0]), jnp.array([0, 0, 1])
-    normal = jnp.where((-0.5 < direction[..., 1]) & (direction[..., 1] < 0.5), y, z)
-    normal = normal - direction * jnp.dot(direction, normal)
+    # x = direction
+    # x = jnp.where(is_zero, jnp.ones_like(x), x)
+    # n = jnp.linalg.norm(x, axis=-1, keepdims=True)
+    # n = jnp.where(is_zero, 0.0, n)
+    # direction = x / n
```
Comment (chungmin99) [inline src/jaxmp/coll/_collide_types.py:42]: Remove the commented lines.

File: src/jaxmp/coll/_collide_types.py [hunk 2]
```diff
@@ -227,7 +247,15 @@ def from_sphere_pairs(sph_0: Sphere, sph_1: Sphere) -> Capsule:
...
 
         radii = sph_0.size[..., 0:1]
-        height = jnp.linalg.norm(sph_1.pos - sph_0.pos, axis=-1, keepdims=True)
+
+        x = sph_1.pos - sph_0.pos
+        is_zero = jnp.allclose(x, 0.0)
+        # temporarily swap x with ones if is_zero, then swap back
+        x = jnp.where(is_zero, jnp.ones_like(x), x)
+        n = jnp.linalg.norm(x, axis=-1, keepdims=True)
+        # n = jnp.where(is_zero, 0.0, n)
```
Comment (chungmin99) [inline src/jaxmp/coll/_collide_types.py:256]: - remove commented line
- explain why the wapping is necessary (e.g., for gradients)

File: src/jaxmp/extras/solve_mpc.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    - Ensure that all constraints are satisfied:
+        - Joint limits
+        - Velocity limits
+        - Collision avoidance
+    """
+    factors: list[jaxls.Factor] = []
+
+    num_targets = target_joint_indices.shape[0]
+
+    def batched_rplus(
```
Comment (chungmin99) [inline src/jaxmp/extras/solve_mpc.py:58]: Add comment to explain why we needed to implement `batched_rplus` (for `jaxls` updates.)

File: src/jaxmp/robot_factors.py [hunk 4]
```diff
@@ -252,50 +266,59 @@ def self_coll_cost(
...
-                        for pair in robot_coll.self_coll_list
-                    ]
-                ),
-                jnp.asarray(
-                    [
-                        robot_coll.link_to_colls[pair[1]]
-                        for pair in robot_coll.self_coll_list
-                    ]
-                ),
+                JointVarType(jnp.repeat(var_idx, num_coll_factors)),
```
Comment (chungmin99) [inline src/jaxmp/robot_factors.py:287]: The code feels messy (with manual batch). Is there a better way to do this?

Comment (chungmin99) [reply src/jaxmp/robot_factors.py:287]: I think we could only fix this if we could vmap across jacobians while setting some `in_axes=None`...? But perhaps we could revisit this in a future PR.
